### PR TITLE
Introduce semi-automated changelog generation

### PR DESCRIPTION
**Pull request by medikoo**
/cc @Mmarzex @Danwakeem @rfaga @hev 

We're leaning in direction of having this project ready for being consumed and promoted among the open-source community.

This PR brings necessary changes to how we handle new releases. It ensures that each release from now on will be properly documented with a changelog, and that it'll happen semi-automatically (assuming we will respect semantic commit messages (as in other open-source projects we maintain):

See document that puts more details into semantic commit message format: https://docs.google.com/document/d/1hKUs3qt_aVp_PBI1UqvfaIqKma3jAJimEoGCRGGbOqs/edit#heading=h.51lj2oiegm91

Having that in, there will be two paths to follow when introducing new changes:
- Propose them through a branch in this repository. Having that count of commits should be minimal (in most cases it should be 1, same as each PR should distinctively handle individual improvement), and commit messages should be clean, in format 1:1 ready for putting into publically available CHANGELOG (formatting of commits will be validated with commitlint)
- Propose changes through a branch in the fork. Then commit messages are whatever works. It's just the developer that will merge the PR should ensure that changes are squash merged with a properly structured commit message,

--- 

- In PR's coming from branches of this repository validate commit messages with `commitlint`
- If PR proposes a new release, validate that changelog for release was provided
- Automatically publish release notes after version publication
- Add documentation on the new release process

### CODE REVIEW

File: .github/workflows/publish-aws-lambda-otel-extension.yml [hunk 1]
```diff
@@ -112,3 +116,10 @@ jobs:
           npm publish
           cd ../aws-lambda-otel-extension-dist
           npm publish
+
+      - name: Publish release notes
+        run: |
+          cd packages/aws-lambda-otel-extension
+          TEMP_ARRAY=($(echo $GITHUB_REF | tr "/" "\n"))
+          TAG=${TEMP_ARRAY[@]: -1}
+          npx github-release-from-cc-changelog $TAG
```
Comment (Danwakeem) [inline .github/workflows/publish-aws-lambda-otel-extension.yml:125]: It looks like this is a package you created 😎 Seems pretty cool. Do we use this in any of our other open source projects?

**Review by Danwakeem — COMMENTED** [review]
Seems reasonable to me 👍 

I am not sure I understand these parts though

- Propose them through a branch in this repository. Having that count of commits should be minimal (in most cases it should be 1, same as each PR should distinctively handle individual improvement), and commit messages should be clean, in format 1:1 ready for putting into publically available CHANGELOG (formatting of commits will be validated with commitlint)
- Propose changes through a branch in the fork. Then commit messages are whatever works. It's just the developer that will merge the PR should ensure that changes are squash merged with a properly structured commit message,

Would we create a fork and then create a PR from the fork?
Or would we crate a branch on this repo then create a PR from the branch?
Or would we just pick either one?

File: .github/workflows/publish-aws-lambda-otel-extension.yml (hunk 1 above)
Comment (medikoo) [reply .github/workflows/publish-aws-lambda-otel-extension.yml:125]: Yes, it's used in the Framework and all related packages

Comment (Danwakeem) [reply .github/workflows/publish-aws-lambda-otel-extension.yml:125]: Very cool! 😎

**Comment by medikoo** [comment]
> Would we create a fork and then create a PR from the fork?
> Or would we crate a branch on this repo then create a PR from the branch?
> Or would we just pick either one?

It's up to you. If you want to practice clean commits in PR, you can go with creating branches in this repository and posting PRs from them.

But if you'd better do whatever commit messages and only at the merge stage decide on a message on one final commit (here it's important that PR covers just one thing and not a few unrelated things), then you should rather go via fork (if PR is from a local branch and commit message won't follow the conventional commit notation CI will not pass)

It's what we practice in the Framework and related packages

**Review by Danwakeem — APPROVED** [review]
😎 Sorry should have approved this one the other day

**Comment by medikoo** [comment]
@Danwakeem @Mmarzex @rfaga As we got this merged. From now versions should not be bumped in functionality PR's.

It's that functionality changes PR's should be merged first with proper semantic commit messages (they will guarantee automated changelog generation that describes changes and points to responsible commits).

And then release PR should follow, which should contain changelog as generated by `npm run prepare-release` command.
See https://github.com/serverless/runtime/blob/main/RELEASE_PROCESS.md for all details on release flow.

**Pull request merged**